From ef299371e17789a28c976e84176ec155d5dc3a93 Mon Sep 17 00:00:00 2001
From: Matt Rogers <mrogers@redhat.com>
Date: Fri, 15 Jul 2016 10:17:45 -0400
Subject: [PATCH 2/2] Add the kprop-port option to kadmind

The -k option for kadmind sets the port number that kprop is spawned
with during an iprop full resync.  Fall back to checking the
KPROP_PORT environment variable if the option is not set.

ticket: 8456 (new)
(cherry picked from commit 7ed18b1e5a11a514461be2281ff884e8173299b1)
[rharwood@redhat.com: nothing is in the same place]
---
 doc/rst_source/krb_admins/admin_commands/kadmind.rst |  7 ++++++-
 src/kadmin/server/ipropd_svc.c                       |  7 +++----
 src/kadmin/server/kadmind.M                          |  7 ++++++-
 src/kadmin/server/ovsec_kadmd.c                      | 11 ++++++++++-
 4 files changed, 25 insertions(+), 7 deletions(-)

diff --git a/doc/rst_source/krb_admins/admin_commands/kadmind.rst b/doc/rst_source/krb_admins/admin_commands/kadmind.rst
index a7ac592..de63f0f 100644
--- a/doc/rst_source/krb_admins/admin_commands/kadmind.rst
+++ b/doc/rst_source/krb_admins/admin_commands/kadmind.rst
@@ -6,7 +6,7 @@ kadmind
 SYNOPSIS
 -----------
        
-**kadmind** [**-x** *db_args*] [**-r** *realm*] [**-m**] [**-nofork**] [**-port** *port-number*] [**-P** *pid_file*]
+**kadmind** [**-x** *db_args*] [**-r** *realm*] [**-m**] [**-nofork**] [**-port** *port-number*] [**-P** *pid_file*] [**-k** *kprop_port*]
 
 DESCRIPTION
 -----------
@@ -87,6 +87,11 @@ OPTIONS
               specifies the file to which the PID of *kadmind* process should be written to after it starts up.  This can be used to identify
               whether *kadmind* is still running and to allow init scripts to stop the correct process.
 
+       **-k** *kprop_port*
+              specifies the port by which the kprop process that is spawned by kadmind
+              connects to the slave kpropd, in order to transfer the dump file during
+              an iprop full resync request.
+
 CONFIGURATION VALUES
 ---------------------------
 
diff --git a/src/kadmin/server/ipropd_svc.c b/src/kadmin/server/ipropd_svc.c
index f8d75d4..5768e6e 100644
--- a/src/kadmin/server/ipropd_svc.c
+++ b/src/kadmin/server/ipropd_svc.c
@@ -35,6 +35,7 @@ extern gss_name_t rqst2name(struct svc_req *rqstp);
 extern void *global_server_handle;
 extern int nofork;
 extern short l_port;
+extern char *kprop_port;
 static char abuf[33];
 
 /* Result is stored in a static buffer and is invalidated by the next call. */
@@ -375,13 +376,11 @@ ipropx_resync(uint32_t vers, struct svc_req *rqstp)
 	    _exit(1);
 	}
 
-
 	DPRINT(("%s: exec `kprop -r %s -f %s %s' ...\n",
 		handle->params.realm, whoami, tmpf, clhost));
-	/* XXX Yuck!  */
-	if (getenv("KPROP_PORT"))
+	if (kprop_port != NULL)
             pret = execl(KPROPD_DEFAULT_KPROP, "kprop", "-r", handle->params.realm, "-f",
-                         tmpf, "-P", getenv("KPROP_PORT"),
+                         tmpf, "-P", kprop_port,
 			 clhost, NULL);
 	else
             pret = execl(KPROPD_DEFAULT_KPROP, "kprop", "-r", handle->params.realm, "-f",
diff --git a/src/kadmin/server/kadmind.M b/src/kadmin/server/kadmind.M
index d54b7c2..1bbbc3e 100644
--- a/src/kadmin/server/kadmind.M
+++ b/src/kadmin/server/kadmind.M
@@ -5,7 +5,7 @@ kadmind \- KADM5 administration server
 .B kadmind
 [\fB\-x\fP \fIdb_args\fP] [\fB-r\fP \fIrealm\fP] [\fB\-m\fP] [\fB\-nofork\fP] [\fB\-port\fP
 \fIport-number\fP]
-    [\fB\-P\fP \fIpid_file\fP]
+    [\fB\-P\fP \fIpid_file\fP] [\fB\-k\fP \fIkprop_port\fP]
 .SH DESCRIPTION
 This command starts the KADM5 administration server.  If the database is db2, 
 the administration server runs on the master Kerberos server, which stores the KDC
@@ -131,6 +131,11 @@ process should be written to after it starts up.  This can be used to
 identify whether
 .B kadmind
 is still running and to allow init scripts to stop the correct process.
+.TP
+\fB\-k\fP \fIkprop_port\fP
+specifies the port by which the kprop process that is spawned by kadmind
+connects to the slave kpropd, in order to transfer the dump file during an
+iprop full resync request.
 .SH CONFIGURATION VALUES
 .PP
 In addition to the relations defined in kdc.conf(5), kadmind
diff --git a/src/kadmin/server/ovsec_kadmd.c b/src/kadmin/server/ovsec_kadmd.c
index f38f209..ac44919 100644
--- a/src/kadmin/server/ovsec_kadmd.c
+++ b/src/kadmin/server/ovsec_kadmd.c
@@ -68,6 +68,7 @@ extern int daemon(int, int);
 gss_name_t gss_changepw_name = NULL, gss_oldchangepw_name = NULL;
 gss_name_t gss_kadmin_name = NULL;
 void *global_server_handle;
+char *kprop_port = NULL;
 
 extern krb5_keylist_node  *master_keylist;
 
@@ -109,7 +110,7 @@ static void usage()
 {
     fprintf(stderr, _("Usage: kadmind [-x db_args]* [-r realm] [-m] [-nofork] "
                       "[-port port-number]\n"
-                      "\t\t[-P pid_file]\n"
+                      "\t\t[-k kprop_port] [-P pid_file]\n"
                       "\nwhere,\n\t[-x db_args]* - any number of database "
                       "specific arguments.\n"
                       "\t\t\tLook at each database documentation for "
@@ -300,6 +301,11 @@ int main(int argc, char *argv[])
             pid_file = *argv;
         } else if (strcmp(*argv, "-W") == 0) {
             strong_random = 0;
+        } else if (strcmp(*argv, "-k") == 0) {
+            argc--; argv++;
+            if (!argc)
+                usage();
+            kprop_port = *argv;
         } else
             break;
         argc--; argv++;
@@ -638,6 +644,9 @@ kterr:
 #endif
     }
 
+    if (kprop_port == NULL)
+        kprop_port = getenv("KPROP_PORT");
+    
     krb5_klog_syslog(LOG_INFO, _("starting"));
     if (nofork)
         fprintf(stderr, _("%s: starting...\n"), whoami);
-- 
2.9.3

